libraries/core_libs/consensus/src/pbft/pbft_manager.cpp

```diff
@@ -990,12 +996,24 @@ void PbftManager::firstFinish_() {
       placeVote(std::move(vote));
     }
   } else {
-    if (auto vote =
-            generateVoteWithWeight(previous_round_next_voted_value_.first, next_vote_type, period, round, step_);
+    // TODO: We should vote for any value that we first saw 2t+1 next votes for in previous round -> in current design
```

Comment:
Seems to me that previous implementation was correct, but that these changes are needed only because of the changes made to behavior of previous_round_next_voted_value_

I think new implementation is a bit more convoluted but should be correct.    Not sure how important it is to actually track which value was next voted first (could differ by node anyway).